src/render/locals.rs

```diff
@@ -89,87 +91,148 @@ pub fn render_locals<'a, 'tcx: 'a>(ecx: &EvalContext<'a, 'tcx>, frame: Option<&F
+        },
+    }
+}
+
+fn pp_value<'a, 'tcx: 'a>(ecx: &EvalContext<'a, 'tcx>, ty: Ty<'tcx>, val: Value) -> EvalResult<'tcx, String> {
+    let layout = ecx.layout_of(ty).map_err(|_| {
+        EvalErrorKind::AssumptionNotHeld // Don't pretty print dst's
+    })?;
+    match ty.sty {
+        TypeVariants::TyRawPtr(TypeAndMut { ty: &TyS { sty: TypeVariants::TyStr, .. }, .. }) |
+        TypeVariants::TyRef(_, &TyS { sty: TypeVariants::TyStr, .. }, _) => {
+            if let Value::ScalarPair(Scalar::Ptr(ptr), Scalar::Bits { bits: len, .. }) = val {
+                if let Ok(allocation) = ecx.memory.get(ptr.alloc_id) {
+                    let offset = ptr.offset.bytes();
+                    if (offset as u128) < allocation.bytes.len() as u128 {
+                        let alloc_bytes = &allocation.bytes[offset as usize.. (offset as usize).checked_add(len as usize).ok_or(EvalErrorKind::AssumptionNotHeld)?];
+                        let s = String::from_utf8_lossy(alloc_bytes);
+                        return Ok(format!("\"{}\"", s));
                     }
                 }
-                Some(&TyS { sty: TypeVariants::TyUint(_), ..}) => {
-                    format!("{0} (0x{0:08X})", bits)
-                }
-                Some(&TyS { sty: TypeVariants::TyInt(_), ..}) => {
-                    format!("{0} (0x{0:08X})", bits as i128)
+            }
+        }
+        TypeVariants::TyAdt(adt_def, _substs) => {
+            if let Value::Scalar(Scalar::Bits { defined: 0, .. }) = val {
+                Err(EvalErrorKind::AssumptionNotHeld)?;
+            }
+
+            let variant = if adt_def.variants.len() == 1 {
+                0
+            } else if let Value::ByRef(_ptr, _align) = val {
+                // FIXME: read the discriminant in a safe way
+                // let ptr = ptr.to_ptr()?;
+                // unsafe { &mut *(ecx as *const EvalContext as *mut EvalContext) }.read_discriminant_as_variant_index(Place::from_ptr(ptr, align), ty)?
+                Err(EvalErrorKind::AssumptionNotHeld)?
+            } else {
+                Err(EvalErrorKind::AssumptionNotHeld)?
+            };
+
+            //println!("{:?} {:?} {:?}", val, ty, adt_def.variants);
+            let mut pretty = format!("{} {{ ", ecx.tcx.absolute_item_path_str(adt_def.did).replace("<", "&lt;").replace(">", "&gt;"));
+            for (i, adt_field) in adt_def.variants[variant].fields.iter().enumerate() {
+                let field_pretty: EvalResult<String> = do catch {
+                    let ValTy { value: field_val, ty: field_ty } = ecx.read_field(val, None, ::rustc::mir::Field::new(i), ty)?;
+                    pp_value(ecx, field_ty, field_val)?
+                };
+                pretty.push_str(&format!("{}: {}, ", adt_field.ident.as_str(), match field_pretty {
+                    Ok(field_pretty) => field_pretty,
+                    Err(_err) => "<span style='color: red;'>&lt;err&gt;</span>".to_string(),
```

Comment:
when does this case occur? if the memory is malformed?

Reply:
Yes, or not yet initialized.